motor-control-firmware/src/encoder.c

```diff
@@ -22,6 +22,9 @@ void encoder_init_primary(void)
     STM32_TIM4->SMCR = STM32_TIM_SMCR_SMS(3); // count on both edges
     STM32_TIM4->CCMR1 = STM32_TIM_CCMR1_CC1S(1); // CC1 channel is input, IC1 is mapped on TI1
     STM32_TIM4->CCMR1 |= STM32_TIM_CCMR1_CC2S(1); // CC2 channel is input, IC2 is mapped on TI2
+    /* enable filtering, clock prescaler 32, 8 samples */
+    STM32_TIM4->CCMR1 |= STM32_TIM_CCMR1_IC1F(0xf);
```

Comment:
What |= do ?

Reply:
`a |= b` means `a = a | b` with `|` being the bitwise OR operator. Here it is used to set some bits:

Suppose you have a variable `foo` which holds the binary value `100` and you want to set the last bit to `1`.
You can do that by doing `foo |= 1;`, here is what the OR will look like:

```
  100
| 001
-----
  101 <- last bit is set
```

Here we use it to set some bits in CCMR1 to 0xf. To know the signification of those bits, you can look in the F3 reference manual, under the chapter for timers.

Reply:
Nice, thank you :)

Reply:
Can you point to examples where there is a lot of preprocessor code ? Usually we use macros as a way to make code more readable. For example, the code above (`STM32_TIM_CCMR1_IC1F(0xf);`) is better than `0xf0` because it allows you to match the value with what you see in the datasheet.

Other than that we often have stuff like `#ifdef` to enable or disable features at compile time. The most common version of that is to build with or without the bootloader enabled, like [here](https://github.com/cvra/robot-software/blob/a5061d965ffeeca5f8466b65f7961a8b3951ac1f/motor-control-firmware/src/bootloader_config.c#L6-L31).

Finally, we sometimes have to write code using the preprocessor rather than simple C functions because there is simply no other way. One example of this is our [logging macros](https://github.com/cvra/robot-software/blob/a5061d965ffeeca5f8466b65f7961a8b3951ac1f/lib/error/error.h#L79-L88): The logging system needs to find the name of the source file and the current line in the file, in order to print those useful informations to the user. Unfortunately, such informations are not possible to find at any other time than preprocessing time: the line number will have changed once pre-preprocessing is finished for example.

Hope that explains a bit why we do it :)